plugin/azure/azure.go

```diff
@@ -0,0 +1,226 @@
+package azure
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	azure "github.com/Azure/azure-sdk-for-go/profiles/latest/dns/mgmt/dns"
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/file"
+	"github.com/coredns/coredns/plugin/pkg/fall"
+	"github.com/coredns/coredns/plugin/pkg/upstream"
+	"github.com/coredns/coredns/request"
+	"github.com/miekg/dns"
+)
+
+type zone struct {
+	id  string
+	z   *file.Zone
+	dns string
+}
+
+type zones map[string][]*zone
+
+// Azure is the core struct of the azure plugin
+type Azure struct {
+	Next      plugin.Handler
+	Fall      fall.F
+	zoneNames []string
+	client    azure.RecordSetsClient
+	upstream  *upstream.Upstream
+	zMu       sync.RWMutex
+	zones     zones
+}
+
+// New validates the input dns zones and initializes the Azure struct
+func New(ctx context.Context, dnsClient azure.RecordSetsClient, keys map[string][]string, up *upstream.Upstream) (*Azure, error) {
+	zones := make(map[string][]*zone, len(keys))
+	zoneNames := make([]string, 0, len(keys))
+	for resourceGroup, inputZoneNames := range keys {
+		for _, zoneName := range inputZoneNames {
+			_, err := dnsClient.ListAllByDNSZone(context.Background(), resourceGroup, zoneName, nil, "")
+			if err != nil {
+				log.Fatalf("err not nil:%s", err.Error())
+			}
+
+			if _, ok := zones[zoneName]; !ok {
+				zoneNames = append(zoneNames, fmt.Sprintf("%s.", zoneName))
```

Comment:
Not very sure about `"%s."`, can you explain a little about the intention here?

Reply:
cmiiw, CoreDNS adds a final trailing dot needed for the root of the DNS tree and makes the `qname` a fqdn. 
In the `zones` field of `Azure` struct, I added the same so that in `ServeDNS`, `plugin.Zones(h.zoneNames).Matches(qname)` results in a successful lookup.

Reply:
I know fqdn with dot, but I am asking about:
```
if _, ok := zones[zoneName]; !ok {
```

I would expect the zoneName is pre-normalized with (or without dot), then having one lookup only?

Reply:
hmm, if I understand correctly, you mean that `zoneName` would already be what it needs to be(be with or without dot `.`) in the `Corefile`? 
This won't work because azure doesn't allow creation of DNS zones with a trailing dot (throws `InvalidResourceNameTrailingCharacters`).
So, the user will always have a DNS zone like `foo.com`, we will fetch the records for `foo.com` using azure sdk but when serving via CoreDNS, it will query for a fqdn and lookup will fail if we don't make this transformation. 

let me know if I misunderstood your question though!